File: pset6/credit/credit.py
from cs50 import get_int


def main():
    # Prompt user for credit card number
    credit_card_number = 0
    while credit_card_number <= 0:
        credit_card_number = get_int("Number: ")

    # Checks if conditions are met
    if check_validity(credit_card_number):
        print_credit_card_brand(credit_card_number)
    else:
        print("INVALID")

# Bool - Check if length and Luhn's algorithm are true


def check_validity(CCN):
    length = find_length(CCN)
    if length == 13 or length == 15 or length == 16 and checksum(CCN):
        return True
    else:
        return False

# Int - Finds out the length of the card


def find_length(CCN):
    length = 0
    while CCN >= 1:
        length += 1
        CCN /= 10
    return length

# Bool - Calculate Luhn's algorithm


def checksum(CCN):
    digit_sum = 0

    for i in range(find_length(CCN)):
        if (i % 2 == 0):
            digit_sum += CCN % 10
        else:
            digit = 2 * (CCN % 10)
            digit_sum += int(digit / 10) + (digit % 10)
        # Takes off last digit of CCN
        CCN -= CCN % 10
        CCN /= 10

    if (digit_sum % 10) == 0:
        return True
    else:
        return False

# Void - Prints the card brand


def print_credit_card_brand(CCN):
    if CCN >= 34e13 and CCN < 35e13 or CCN >= 37e13 and CCN <= 38e13:
        print("AMEX")
    elif CCN >= 51e14 and CCN < 56e14:
        print("MASTERCARD")
    elif CCN >= 4e12 and CCN < 5e12 or CCN >= 4e15 and CCN < 5e15:
        print("VISA")
    else:
        print("INVALID")


main()

File: pset1/credit/credit.c
#include <cs50.h>
#include <stdio.h>


bool check_validity(long long CCN);
int find_length(long long CCN);
bool checksum(long long CCN);
void print_credit_card_brand(long long CCN);


int main(void)
{
    //Get credit card number from user
    long long credit_card_number; 
    do
    {
        credit_card_number = get_long_long("Number: ");
    }
    while (credit_card_number < 0);
    
    //Checks if conditions are met
    if (check_validity(credit_card_number)) 
    {
        print_credit_card_brand(credit_card_number);
    }
    else
    {
        printf("INVALID\n");
    }
}


//Check if length and Luhn's algorithm are true
bool check_validity(long long CCN)
{
    int len = find_length(CCN);
    
    return (len == 13 || len == 15 || len == 16) && checksum(CCN);
}


//Finds out the length of the card
int find_length(long long CCN)
{
    int len;
    for (len = 0; CCN != 0; CCN /= 10, len++);
    return len;
}

//Calculate Luhn's algorithm
bool checksum(long long CCN)
{
    int sum = 0;
    for (int i = 0; CCN != 0; i++, CCN /= 10)
    {
        if (i % 2 == 0)
        {
            sum = sum + CCN % 10;
        }
        else 
        {
            int digit = 2 * (CCN % 10);
            sum = sum + digit / 10 + digit % 10;
        }
    }
    return (sum % 10) == 0;
}

//Finds out the card's brand
void print_credit_card_brand(long long CCN)
{
    if ((CCN >= 34e13 && CCN < 35e13) || (CCN >= 37e13 && CCN < 38e13))
    {
        printf("AMEX\n");
    }
    else if (CCN >= 51e14 && CCN < 56e14)
    {
        printf("MASTERCARD\n");
    }
    else if ((CCN >= 4e12 && CCN < 5e12) || (CCN >= 4e15 && CCN < 5e15))
    {
        printf("VISA\n");
    }
    else
    {
        printf("INVALID\n");
    }
}